app/wasm/message_plugin.go

```diff
@@ -40,52 +44,70 @@ func (m *CustomMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddre
 			return nil, nil, sdkerrors.Wrap(err, "osmosis msg")
 		}
-		// if contractMsg.MintTokens != nil {
-		// 	return m.mintTokens(ctx, contractAddr, contractMsg.MintTokens)
-		// }
+		if contractMsg.MintTokens != nil {
+			return m.mintTokens(ctx, contractAddr, contractMsg.MintTokens)
+		}
 		if contractMsg.Swap != nil {
 			return m.swapTokens(ctx, contractAddr, contractMsg.Swap)
 		}
 	}
 	return m.wrapped.DispatchMsg(ctx, contractAddr, contractIBCPortID, msg)
 }
 
-// func (m *CustomMessenger) mintTokens(ctx sdk.Context, contractAddr sdk.AccAddress, mint *wasmbindings.MintTokens) ([]sdk.Event, [][]byte, error) {
-// 	err := PerformMint(m.bank, ctx, contractAddr, mint)
-// 	if err != nil {
-// 		return nil, nil, sdkerrors.Wrap(err, "perform mint")
-// 	}
-// 	return nil, nil, nil
-// }
-
-// func PerformMint(b *bankkeeper.BaseKeeper, ctx sdk.Context, contractAddr sdk.AccAddress, mint *wasmbindings.MintTokens) error {
-// 	if mint == nil {
-// 		return wasmvmtypes.InvalidRequest{Err: "mint token null mint"}
-// 	}
-// 	rcpt, err := parseAddress(mint.Recipient)
-// 	if err != nil {
-// 		return err
-// 	}
-
-// 	denom, err := GetFullDenom(contractAddr.String(), mint.SubDenom)
-// 	if err != nil {
-// 		return sdkerrors.Wrap(err, "mint token denom")
-// 	}
-// 	if mint.Amount.IsNegative() {
-// 		return wasmvmtypes.InvalidRequest{Err: "mint token negative amount"}
-// 	}
-// 	coins := []sdk.Coin{sdk.NewCoin(denom, mint.Amount)}
-
-// 	err = b.MintCoins(ctx, gammtypes.ModuleName, coins)
-// 	if err != nil {
-// 		return sdkerrors.Wrap(err, "minting coins from message")
-// 	}
-// 	err = b.SendCoinsFromModuleToAccount(ctx, gammtypes.ModuleName, rcpt, coins)
-// 	if err != nil {
-// 		return sdkerrors.Wrap(err, "sending newly minted coins from message")
-// 	}
-// 	return nil
-// }
+func (m *CustomMessenger) mintTokens(ctx sdk.Context, contractAddr sdk.AccAddress, mint *wasmbindings.MintTokens) ([]sdk.Event, [][]byte, error) {
+	err := PerformMint(m.tokenFactory, m.bank, ctx, contractAddr, mint)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "perform mint")
+	}
+	return nil, nil, nil
+}
+
+func PerformMint(f *tokenfactorykeeper.Keeper, b *bankkeeper.BaseKeeper, ctx sdk.Context, contractAddr sdk.AccAddress, mint *wasmbindings.MintTokens) error {
+	if mint == nil {
+		return wasmvmtypes.InvalidRequest{Err: "mint token null mint"}
+	}
+	rcpt, err := parseAddress(mint.Recipient)
+	if err != nil {
+		return err
+	}
+
+	// Check if denom is valid
+	denom, err := GetFullDenom(contractAddr.String(), mint.SubDenom)
+	if err != nil {
+		return err
+	}
+
+	if mint.Amount.IsZero() {
+		return wasmvmtypes.InvalidRequest{Err: "mint token zero amount"}
+	}
+	if mint.Amount.IsNegative() {
+		return wasmvmtypes.InvalidRequest{Err: "mint token negative amount"}
+	}
+	coin := sdk.NewCoin(denom, mint.Amount)
+
+	msgServer := tokenfactorykeeper.NewMsgServerImpl(*f)
+
+	// Check if denom already exists
```

Comment:
Okay, this is the critical question.

Assuming we never set any metadata that is not present in the MintTokens command, this works great and makes a simple API.

However, if they plan to add more metadata there, we will want a separate command to run once to register the metadata (and another to query it?).

Reply:
I would merge this as is (we can test stuff with it), but have a discussion with @ValarDragon and @sunnya97 before tagging an Osmosis release.

If they want to add more required metadata, they should add it in Go and we should expose it in Rust now. Otherwise, it will most likely be a breaking contract API change, which we want very much to avoid.